src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,306 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+                //   }
+                //   staticBase = fastPath;
+                //
+
+                // The initialization check looks like this for JIT:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  AND       int
+                //    |  +--*  IND       int
+                //    |  |  \--*  CNS_INT(h) long   0x.... const ptr
+                //    |  \--*  CNS_INT   int    1 (bit mask)
+                //    \--*  CNS_INT   int    1
+                //
+                // For NativeAOT it's:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  IND       nint
+                //    |  \--*  ADD       long
+                //    |     +--*  CNS_INT(h) long   0x.... const ptr
+                //    |     \--*  CNS_INT   int    -8 (offset)
+                //    \--*  CNS_INT   int    0
+                //
+                assert(flagAddr.accessType == IAT_VALUE);
+
+                GenTree* cachedStaticBase = nullptr;
+                GenTree* isInitAdrNode;
+                GenTree* isInitedValue;
+                if (IsTargetAbi(CORINFO_NATIVEAOT_ABI))
+                {
+                    assert(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+                    GenTree* baseAddr = gtNewIconHandleNode((size_t)flagAddr.addr, GTF_ICON_GLOBAL_PTR);
+
+                    // Save it to a temp - we'll be using its value for the replacementNode.
+                    // This leads to some size savings on NativeAOT
+                    if ((staticBaseAddr.addr == flagAddr.addr) && (staticBaseAddr.accessType == flagAddr.accessType))
+                    {
+                        cachedStaticBase = fgInsertCommaFormTemp(&baseAddr);
+                    }
+
+                    // Don't fold ADD(CNS1, CNS2) here since the result won't be reloc-friendly for AOT
+                    isInitAdrNode = gtNewIndir(TYP_I_IMPL, gtNewOperNode(GT_ADD, TYP_I_IMPL, baseAddr,
+                                                                         gtNewIconNode(isInitOffset)));
+                    // 0 means "initialized" on NativeAOT
+                    isInitedValue = gtNewIconNode(0, TYP_I_IMPL);
+                }
+                else
+                {
+                    isInitAdrNode =
+                        gtNewIndOfIconHandleNode(TYP_INT, (size_t)flagAddr.addr, GTF_ICON_GLOBAL_PTR, false);
+
+                    // Check ClassInitFlags::INITIALIZED_FLAG bit
+                    isInitAdrNode = gtNewOperNode(GT_AND, TYP_INT, isInitAdrNode, gtNewIconNode(1));
+                    isInitedValue = gtNewIconNode(1);
+                }
+
+                // This indir points to a mutable location and doesn't have side-effects
+                isInitAdrNode->gtFlags &= ~GTF_EXCEPT;
+                isInitAdrNode->gtFlags |= (GTF_IND_NONFAULTING | GTF_GLOB_REF);
```

Comment:
This looks like it will apply `GTF_IND_NONFAULTING` to `GT_AND` in the Jit case.

Instead of this flag manipulation, it is better to pass `GTF_IND_NONFAULTING` directly to `gtNewIndir` (`GTF_GLOB_REF` still needs to be set manually for now).

Reply:
Thanks! Seems like no need to propagate GLOB_REF to AND - it will be propagated by updatesideeffects later:
```
N007 ( 11, 18) [000014] ----G------                         *  JTRUE     void  
N006 (  9, 16) [000013] J---G------                         \--*  EQ        int   
N004 (  7, 14) [000011] ----G------                            +--*  AND       int   
N002 (  5, 12) [000009] n---G------                            |  +--*  IND       int   
N001 (  3, 10) [000008] H----------                            |  |  \--*  CNS_INT(h) long   0xd1ffab1e global ptr
N003 (  1,  1) [000010] -----------                            |  \--*  CNS_INT   int    1
N005 (  1,  1) [000012] -----------                            \--*  CNS_INT   int    1
```